### ISSUE TITLE
[redux] resource list read request only processes one

### ISSUE THREAD

**Issue by Geczy** [diegohaz/arc#303]
I see two read requests for both of my dispatches, but it only ever finishes with one

Is this a race condition? 

```js
const mapDispatchToProps = dispatch => ({
  readList: () => dispatch(resourceListReadRequest('posts')),
  readStats: () => dispatch(resourceListReadRequest('comments')),
})
```

```js
  componentWillMount() {
    this.props.readList()
    this.props.readStats()
  }
```

You can recreate by cloning redux branch, and adding the readStats() dispatch I have above to here
https://github.com/diegohaz/arc/blob/redux/src-example/containers/PostList.js#L40

![image](https://user-images.githubusercontent.com/1036968/27964121-2b660522-6305-11e7-8772-fe2d23e77a5a.png)

**Comment by Geczy** [comment]
I can solve the problem by doing this

```js
  componentWillMount() {
    this.props.readList().then(() => {
      this.props.readStats()
    })
  }
```

But I don't understand why the first bug is happening?

**Comment by diegohaz** [comment]
That's strange. Will take a look at this later today.

**Comment by mrotaru** [comment]
We're using Arc at work, and ran into this as well. This issue occurs only when the dispatched actions are of the same type. Long story short, I think `fork` can be used instead of `call` to fix this. I'm still trying to wrap my head around `redux-saga` & co, so take the below with a grain of salt.

Taking this code as an example:
```js
export function* readResourceList(api, { params }, { resource, thunk }) {
  try {
    const list = yield call([api, api.get], `/${resource}`, { params })
    yield put(actions.resourceListReadSuccess(resource, list, { params }, thunk))
  } catch (e) {
    yield put(actions.resourceListReadFailure(resource, e, { params }, thunk))
  }
}

export function* watchResourceListReadRequest(api) {
  while (true) {
    const { payload, meta } = yield take(actions.RESOURCE_LIST_READ_REQUEST)
    yield call(readResourceList, api, payload, meta)
  }
}
```

The watchers will be "forked off" before any action is dispatched.  So `watchResourceListReadRequest` runs; `take` will register a listener for any `RESOURCE_LIST_READ_REQUEST` actions that will be dispatched in the future, and the generator stops.

When our component mounts,  the first `RESOURCE_LIST_READ_REQUEST` action is dispatched. The `take` listener will be triggered, "taking" the action from the channel and the watcher generator function will proceed to the next `yield`. 

This is when `readResourceList` is called. Because it returns a generator, `call` will pause the watcher generator until the `readResourceList` generator completes. But before it does, we can dispatch actions of the same type. Redux saga will see them, and check to see if there are any registered "takers". There aren't, because once an action is `take`n, the listener is removed.

Essentially, we need the watcher generator to advance, so that the "while" block is ran again, and 
a new "taker" is registered when `take` is called. But, because of how `call` works when the called function returns a generator, it will wait for it to finish first, and only then will advance the generator. This is why actions are 'ignored' - there are no takers.

To avoid this blocking, we can use `fork` - because it will advance the watcher generator immediately after calling `readResourceList`, without waiting for it to complete.

**Comment by orther** [comment]
Handling every incoming action and spawning a new task saga is very common in redux-saga. To help you write more declarative code they provide the `takeEvery` helper effect to handle this. `takeEvery` essentially does what you described under the hood (as you can [here](https://redux-saga.js.org/docs/advanced/Concurrency.html#takeevery)). 

Below is a quick (not tested) refactor of your above code. I left the destructuring of the action out of the arguments and on multiple lines to make sure it was easy to understand what was passed in.

```js
export function* readResourceList(api, action) {
  const { payload, meta } = action;
  const { params } = payload;
  const { resource, thunk } = meta;

  try {
    const list = yield call([api, api.get], `/${resource}`, { params });
    yield put(
      actions.resourceListReadSuccess(resource, list, { params }, thunk)
    );
  } catch (e) {
    yield put(actions.resourceListReadFailure(resource, e, { params }, thunk));
  }
}

export function* watchResourceListReadRequest(api) {
  yield takeEvery(actions.RESOURCE_LIST_READ_REQUEST, readResourceList, api);
}
```

The above example is to demonstrate the `takeEvery` replacing the actual while loop code in it's place. I would refactor this a bit for better separation of concerns, decoupling, etc. Here the watcher saga's responsibility is to handle the action data and call the read saga.
```js
export function* readResourceList(api, { params }, { resource, thunk }) {
  try {
    const list = yield call([api, api.get], `/${resource}`, { params });
    yield put(actions.resourceListReadSuccess(resource, list, { params }, thunk));
  } catch (e) {
    yield put(actions.resourceListReadFailure(resource, e, { params }, thunk));
  }
}

export function* watchResourceListReadRequest(api, { payload, meta }) {
  yield call(readResourceList, api, payload, meta);
}

export default function* rootSaga() {
  yield takeEvery(actions.RESOURCE_LIST_READ_REQUEST, watchResourceListReadRequest, api);
  // ...
}
```

**BEWARE!** I wrote this code in the github issue editor, so it may have typos/bugs/...  😨 But even if it does, the general concept should still make it through.

Here are two redux-saga doc pages related to this declarative "push" style action handling:
https://redux-saga.js.org/docs/basics/UsingSagaHelpers.html
https://redux-saga.js.org/docs/basics/DeclarativeEffects.html

**Comment by diegohaz** [comment]
@mrotaru and @orther,

You guys rock!

**Comment by diegohaz** [comment]
Fixed in #306
